python/mxnet/symbol/symbol.py

```diff
@@ -1554,7 +1554,37 @@ def optimize_for(self, backend, args=None, aux=None, ctx=None, **kwargs):
                     aux[idx] = NDArray(NDArrayHandle(new_aux_handle[i]))
                 else:
                     aux.append(NDArray(NDArrayHandle(new_aux_handle[i])))
-        return Symbol(out)
+
+        new_sym = Symbol(out)
+
+        if args is not None:
+            new_arg_names = new_sym.list_arguments()
+            deleted_arg_names = set([item for item in arg_names
+                                     if item not in set(new_arg_names)])
+            if isinstance(args, dict):
+                for a_n in deleted_arg_names:
+                    if a_n in args:
+                        args.pop(a_n)
+            elif isinstance(args, list):
+                indices_to_delete = [i for i, name in enumerate(arg_names) if name in deleted_arg_names]
+                for idx in reversed(indices_to_delete):
+                    args.pop(idx)
+
+        if aux is not None:
+            new_aux_names = new_sym.list_auxiliary_states()
+            deleted_aux_names = set([item for item in aux_names
+                                     if item not in set(new_aux_names)])
+
+            if isinstance(aux, dict):
+                for a_n in deleted_aux_names:
+                    if a_n in aux:
+                        aux.pop(a_n)
+            elif isinstance(args, list):
+                indices_to_delete = [i for i, name in enumerate(aux_names) if name in deleted_aux_names]
+                for idx in reversed(indices_to_delete):
+                    aux.pop(idx)
```

Comment:
@ptrendx I remember we discovered a problem previously that the ordering of inputs may change after partitioning #15362. What should we do here if the args/aux are a list and not a dict? Should we re-order them so they match the new ordering after calling optimize_for (which can execute graph partitioning, or graph passes)?

Reply:
The API should not accept lists - it is a source of a lot of evil.

Reply:
Also, @mk-61 FYI.

Reply:
PR removing support for lists as args/aux https://github.com/apache/incubator-mxnet/pull/18483